wasm module for signing client side signing API requests

peeech opened the pull request.

peeech committed:
Initial client commit

peeech committed:
server: update description

peeech committed:
client: create struct HpAdminKeypair

peeech committed:
client: add tests

peeech committed:
add wasm_bindgen

peeech committed:
add inline docs

peeech committed:
client: require strict formatting of payload for signing

peeech committed:
client: add wasm_bindgen_test

peeech committed:
update READMEs

peeech committed:
client: fix tests/www/

peeech committed:
client: add docs

peeech committed:
client: add license

lukateras commented on client/Cargo.toml:
```diff
@@ -0,0 +1,21 @@
+wasm-bindgen-cli = "0.2.51"
+
+[lib]
+crate-type = ["cdylib", "rlib"]
```
Missing EOL. Set up auto EOL in your editor.

lukateras commented on server/Cargo.toml:
```diff
@@ -1,8 +1,10 @@
 [package]
 name = "hp-admin-crypto-server"
 version = "0.1.0"
+description = "Servis for checking authorizatoin of API calls from HP Admin UI"
```
```suggestion
description = "Service for checking authorization of API calls from HP Admin UI"
```

lukateras reviewed (commented):
I'll do full review later today.

lukateras commented on .gitignore:
```diff
@@ -1,2 +1,5 @@
 target/
+pkg/
+node_modules/
+/client/tests/www/node_modules/
```
```suggestion
```

peeech committed:
code cleanup

lukateras committed:
client: run `cargo fmt`

mjbrisebois commented on client/Cargo.toml:
```diff
@@ -0,0 +1,25 @@
+[package]
+name = "hp-admin-crypto-client"
```
```suggestion
name = "@holo-host/hp-admin-key-manager"
```

lukateras replied:
TBH I'd rather standardize on `Keypair` so that we're consistent with `ed25519_dalek`. Also, scope is set using `--scope` flag rather than here I think. Additionally, this might end up including more than just Keypair wrapper.

peeech replied:
I like that name. Accepted.

mjbrisebois commented on client/docs/hp_admin_crypto_client.js.html:
```diff
@@ -0,0 +1,247 @@
+<!DOCTYPE html>
```
I don't think we should have generated documentation in the main branches (eg. master, develop).  `gh-pages` is a good branch to hold these files so they don't clutter the review process

lukateras commented on client/CONTRIBUTING.md:
````diff
@@ -0,0 +1,43 @@
+
+## Generate docs
+```
+./node_modules/.bin/jsdoc pkg/hp_admin_crypto_client.js --configure .jsdoc.json --destination docs
````
```suggestion
npx jsdoc pkg/hp_admin_crypto_client.js --configure .jsdoc.json --destination docs
```

mjbrisebois replied:
There is no `jsdoc` in this repo.  It's hard to say where it belongs because the only `package.json` file is in `client/tests/www/`.  I suggest moving that `package.json` into `client/` and then adding `jsdoc` as a dev dependency.

lukateras commented on client/CONTRIBUTING.md:
```diff
@@ -0,0 +1,43 @@
+`nix-shell` sets the environment for development and tests. Type `nix-shell` from project's root directory to set up dev environment, then cd to `client` folder.
+
+## Build
+Build from source with `wasm-pack build`.
```
```suggestion
Build from source with `wasm-pack build --scope holo-host`.
```

lukateras commented on client/README.md:
```diff
@@ -0,0 +1,29 @@
+
+const HC_PUBLIC_KEY = "3llrdmlase6xwo9drzs6qpze40hgaucyf7g8xpjze6dz32s957";
+const EMAIL= "pj@abba.pl";
+const PASSWORD= "abba";
```
```suggestion
const PASSWORD = "abba";
```

lukateras commented on client/README.md:
```diff
@@ -0,0 +1,29 @@
+import {HpAdminKeypair} from "hp-admin-crypto-client";
+
+const HC_PUBLIC_KEY = "3llrdmlase6xwo9drzs6qpze40hgaucyf7g8xpjze6dz32s957";
+const EMAIL= "pj@abba.pl";
```
```suggestion
const EMAIL = "pj@abba.pl";
```

mjbrisebois commented on server/README.md:
```diff
@@ -0,0 +1,30 @@
+# HoloPort Admin Crypto Client
```
```suggestion
# HoloPort Admin Crypto Server
```

lukateras commented on client/docs/hp_admin_crypto_client.js.html:
```diff
@@ -0,0 +1,247 @@
+<!DOCTYPE html>
```
Maybe I could set up Hydra (actual Hydra, not Nginx) to build and continuously deliver docs instead?

peeech replied:
That would be great. How can we contain the information about CD process in this repo?

lukateras commented on client/src/lib.rs:
```diff
@@ -0,0 +1,177 @@
+struct Payload {
+    method: String,
+    uri: String,
+    body_string: String,
```
```suggestion
    body: String,
```

Seems to be asymmetrical  to `method` and `uri` (why not `method_string` and `uri_string`?).

mjbrisebois replied:
We should match the RFC which says `method`, `request` and `body`

peeech replied:
will do

lukateras commented on client/src/lib.rs:
```diff
@@ -0,0 +1,177 @@
+        password: String,
+    ) -> Result<HpAdminKeypair, JsValue> {
+        // rewrite any Error into JsValue
+        match new_inner(hc_public_key_string, email, password) {
```
I know some of our older WASM code uses inner/outer split to handle errors properly. Have you tried `into_js_error` function, as seen in Chaperone `KeyManager`? I think it'd be much easier to read.

peeech replied:
leaving as is

lukateras commented on client/src/lib.rs:
```diff
@@ -0,0 +1,177 @@
+        let payload_vec = match parse_payload(payload) {
+            Ok(v) => v,
+            Err(_) => {
+                return Err("Malformed signing payload, check docs for details."
```
Errors that are ignored here produce much better/more detailed error messages than this.  recommend to just return internal Serde error instead of using custom handling/text.

lukateras commented on client/tests/www/index.js:
```diff
@@ -0,0 +1,15 @@
+    body_string: ""
+}
+
+console.log(kp.sign(payload));
```
EOL.

lukateras commented on client/README.md:
````diff
@@ -0,0 +1,29 @@
+
+## Usage
+```javascript
+import {HpAdminKeypair} from "hp-admin-crypto-client";
````
```suggestion
import { HpAdminKeypair } from "@holo-host/hp-admin-crypto-client";
```

lukateras commented on client/src/lib.rs:
```diff
@@ -0,0 +1,177 @@
+        let signature = self.0.sign(&payload_vec);
+        Ok(base64::encode_config(
+            &signature.to_bytes()[..],
+            base64::STANDARD_NO_PAD,
```
TBH even though I've started this trend, I think it would be best to standardize on, well, standard Base64.

mjbrisebois commented on client/CONTRIBUTING.md:
```diff
@@ -0,0 +1,43 @@
+`nix-shell` sets the environment for development and tests. Type `nix-shell` from project's root directory to set up dev environment, then cd to `client` folder.
+
+## Build
+Build from source with `wasm-pack build`.
```
FYI: We run `wasm-pack build` twice so that we have an implementation that works for both node and web.  It makes usage testing in nodejs easy to implement which doubles as validated user documentation.

See [build.sh](https://github.com/Holo-Host/chaperone-key-manager/blob/develop/build.sh) from Chaperone Key Manager.

peeech replied:
Implemented by Perry

mjbrisebois reviewed (changes requested):
Only found (suggested) some minor changes.  Tests and build worked first try.

pjkundert committed:
Update to latest holo-nixpkgs
o Make into a Cargo workspace with projects
o Update to hpos-config from hpos-state

pjkundert committed:
Further Makefile targets for building server, client/pkg, running test

peeech committed:
Merge pull request #5 from Holo-Host/201912-review-cleanup

201912 review cleanup

peeech committed:
hp-admin-crypto-client -> hp-admin-key-manager, remove docs

peeech committed:
client: add gh-pages link to docs in README.md

pjkundert committed:
Correct parsing of `curl` output, correct expected failing server test

lukateras commented on Makefile:
```diff
@@ -48,10 +48,11 @@ crypto-server.PID:
 	    http://127.0.0.1:2884 2>&1 \
-	  | sed -ne "s/< HTTP\/1.1 \(.*\)/\1/p" ); \
-	[[ "$$GOT" == "200 OK" ]] && echo OK || echo "ERROR: '200 OK' Expected; got '$$GOT'"; \
+	  | sed -ne "s/^< HTTP\/1.1 \([^\r\n]*\).*$$/\1/p" ); \
```
🤪

pjkundert committed:
Attemp to use js-sys for JsValue Err, and console_error_panic_hook
o Loses any kind of diagnostic descriptive error on bad payload
o This doesn't appear to be an improvement, to me...

mjbrisebois commented on client/README.md:
```diff
@@ -0,0 +1,29 @@
+    method: "get",
+    uri: "/someuri",
+    body_string: ""
+}
```
RFC https://github.com/Holo-Host/rfcs/blob/master/2019-11-03-hp-dispatcher/README.md#x-hpos-admin-signature defines structure as `method`, `request` and `body`

mjbrisebois commented on client/CONTRIBUTING.md:
````diff
@@ -0,0 +1,44 @@
+## Generate docs
+Publish docs only on `gh-pages` branch with:
+```
+npx jsdoc pkg/hp-admin-key-manager.js --configure .jsdoc.json --destination docs
````
`jsdoc` is still missing in the repo.  `package-lock.json` was committed but it looks like `package.json` was forgotten

mjbrisebois commented on client/Cargo.toml:
```diff
@@ -0,0 +1,25 @@
+[package]
+name = "hp-admin-key-manager"
```
I'm in favor of calling this `hp-admin-keypair` if everyone else likes that better

peeech replied:
done

peeech committed:
rename body_string -> body in payload

peeech committed:
client: fix wasm-pack tests

peeech committed:
Merge pull request #6 from Holo-Host/201912-use-js-sys

Use js-sys for JsValue Err, and console_error_panic_hook

peeech committed:
payload.uri -> payload.request

peeech committed:
ensure client signature matches server

peeech committed:
Merge branch '201912-use-js-sys' into 2019-12-12-crypto-client

peeech committed:
hp-admin-key-manager -> hp-admin-keypair

peeech committed:
update client/tests/www

peeech committed:
update test to lates npm version

peeech committed:
run `cargo fmt`

mjbrisebois reviewed (approved):
Looks ready to me.  RFC [PR#22](https://github.com/Holo-Host/rfcs/pull/22) may cause more changes depending on the conclusion.

The only remaining change requests I have would be polishing `jsdoc` documentation and adding usage (nodejs or puppeteer) tests

lukateras commented on client/src/lib.rs:
```diff
@@ -0,0 +1,189 @@
+    }
+}
+
+fn new_inner(hc_public_key_string: String, email: String, password: String) -> Fallible<Keypair> {
```
`new_inner` still doesn't use `into_js_error`, which should've been addressed as part of #6.

lukateras commented on Makefile:
```diff
@@ -0,0 +1,58 @@
+# you add `substituters = ...` and `trusted-public-keys = ...` to your nix.conf (see README.md), to
+# take advantage of cached Nix and Holo build assets.
+nix-%:
+	nix-shell --pure --run "make $*"
```
This is a bad idea, please don't add this to Makefile. Users should enter `nix-shell`, rather than creating and destroying shell environment on every command. This can have all sorts of unintended consequences and bad user experience, ranging from services being started/stopped on every command, to build times that are not adequate for iterative development (usually entry takes at least 5 sec on commodity hardware). This is not how Nix shell is supposed to be used. I've stated this before in one of `happ-example` reviews.

lukateras commented on Makefile:
```diff
@@ -0,0 +1,58 @@
+	    http://127.0.0.1:2884 2>&1 \
+	  | sed -ne "s/^< HTTP\/1.1 \([^\r\n]*\).*$$/\1/p" ); \
+	[[ "$$GOT" == "$$EXP" ]] && printf "OK: Expected: '%s'" "$$EXP" || printf "*** ERROR: Expected: '%s', Got: '%s'" "$$EXP" "$$GOT"; \
+	kill $$PID
```
This is not readable or maintainable.

pjkundert committed:
Remove unused target from Makefile

peeech committed:
convert failure::error to JsValue
